Allow multiple SubPlan references to a single subplan.

hlinnaka opened the pull request:
This is a refactoring of the planner post-processing steps. I started out doing this while working on the 9.6 merge branch, but it grew into this much bigger ball of effort, and it seemed best to put this in a separate PR.

Aside from moving code around in cdbllize.c and cdbmutate.c, the big change here is that we no longer duplicate subplans, when there are multiple SubPlan references to the same subplan. That mechanism, (`fixup_subplans()`) was added in commit d0aea18475, but it seems better to tolerate multiple SubPlan references.

hlinnaka commented:
See also discussion on gpdb-dev, on some bigger plans I have in this area: https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/JqCFUCQeqQU

hlinnaka commented:
Note: I opened a separate PR about the first commit in this series earlier already: https://github.com/greenplum-db/gpdb/pull/8612. The VALUES fixing commit is another independent change that I opened a separate PR for: https://github.com/greenplum-db/gpdb/pull/8623. But it's needed for the rest of this PR to work correctly.

bhuvnesh2703 commented on src/backend/cdb/cdbllize.c:
```diff
@@ -770,17 +864,11 @@ ParallelizeSubplan(SubPlan *spExpr, PlanProfile *context)
-	 * Replace subplan in global subplan list.
-	 */
-	list_nth_replace(context->root->glob->subplans, spExpr->plan_id - 1, newPlan);
-	spExpr->is_parallelized = true;
```
I think your refactoring is great.. 
the existing code assumes that we will come to parallelize the same subplan and in that case we shouldn't parallelize as it maintains the is_parallelized flag in SubPlan node. However, even though if two subplans refer to the same plan_id it does not consider that two different addresses of subplan refer to the same plan and waas one of the reason why we ran into issue of parallelizing the same plan_id referred by them twice / more and ran into the issue.
Consider the case where the predicate is
`var1 = subplan` and `subplan = var2`
even though these subplan may be referring same plan_id but they could be different node.

I tried the queries included with the earlier fix, they all work now without the duplication as you are maintaining PlanProfileSubPlanInfo based on the plans available in root->glob->subplans not in the subplan node. So if the plan in root->glob->subplans is seen once you skip parallelizing it again. Thats good and ideally what should have been done with the fixup_subplans fix earlier. 

One of the queries which was added with the commit: d0aea1847518ce86dd0774d3ea20ab967bdba879 produced the plan as below. Two subplans are used once in Hash Cond `Hash Cond: public.foo_s.b = ((subplan))` and other in `Filter: c = ((subplan))`. But now if we see this plan, Hash Cond refers to `Hash Cond: (foo_s_1_prt_sub_one.b = bar_s.c)` where bar_s.c is propagated from the inner child of the HJ so even though in difference slices we don't run into the issue. correct me if wrong.
```sql
bhuvneshchaudhary=# explain SELECT bar_s.c FROM bar_s, foo_s WHERE foo_s.b = (SELECT max(i) FROM baz_s WHERE bar_s.c = 9) AND bar_s.c::int4 = (SELECT max(i) FROM baz_s WHERE bar_s.c = 9);
                                                              QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 2:1  (slice4; segments: 2)  (cost=2.12..2721.41 rows=173 width=4)
   ->  Hash Join  (cost=2.12..2721.41 rows=87 width=4)
         Hash Cond: public.foo_s.b = ((subplan))
         ->  Append  (cost=0.00..1922.00 rows=86100 width=4)
               ->  Seq Scan on foo_s_1_prt_sub_one foo_s  (cost=0.00..961.00 rows=43050 width=4)
               ->  Seq Scan on foo_s_1_prt_sub_two foo_s  (cost=0.00..961.00 rows=43050 width=4)
         ->  Hash  (cost=2.10..2.10 rows=1 width=4)
               ->  Broadcast Motion 2:2  (slice2; segments: 2)  (cost=0.00..2.10 rows=1 width=4)
                     ->  Seq Scan on bar_s  (cost=0.00..2.07 rows=1 width=4)
                           Filter: c = ((subplan))
                           SubPlan 2
                             ->  Aggregate  (cost=1.05..1.06 rows=1 width=4)
                                   ->  Result  (cost=0.00..1.01 rows=1 width=4)
                                         One-Time Filter: $0 = 9
                                         ->  Result  (cost=1.01..1.02 rows=1 width=4)
                                               ->  Materialize  (cost=1.01..1.02 rows=1 width=4)
                                                     ->  Broadcast Motion 2:2  (slice1; segments: 2)  (cost=0.00..1.01 rows=1 width=4)
                                                           ->  Seq Scan on baz_s  (cost=0.00..1.01 rows=1 width=4)
         SubPlan 1
           ->  Aggregate  (cost=1.05..1.06 rows=1 width=4)
                 ->  Result  (cost=0.00..1.01 rows=1 width=4)
                       One-Time Filter: $0 = 9
                       ->  Result  (cost=1.01..1.02 rows=1 width=4)
                             ->  Materialize  (cost=1.01..1.02 rows=1 width=4)
                                   ->  Broadcast Motion 2:2  (slice3; segments: 2)  (cost=0.00..1.01 rows=1 width=4)
                                         ->  Seq Scan on baz_s  (cost=0.00..1.01 rows=1 width=4)
 Optimizer status: legacy query optimizer
```

Now the plan is `Hash Cond: (foo_s_1_prt_sub_one.b = bar_s.c)`
```sql
bhuvneshchaudhary=# explain SELECT bar_s.c FROM bar_s, foo_s WHERE foo_s.b = (SELECT max(i) FROM baz_s WHERE bar_s.c = 9) AND foo_s.b = bar_s.c::int4;
                                                                 QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 2:1  (slice3; segments: 2)  (cost=2.12..2786.84 rows=173 width=4)
   ->  Hash Join  (cost=2.12..2786.84 rows=87 width=4)
         Hash Cond: (foo_s_1_prt_sub_one.b = bar_s.c)
         ->  Append  (cost=0.00..1922.00 rows=86100 width=4)
               ->  Result  (cost=0.00..961.00 rows=43050 width=4)
                     One-Time Filter: PartSelected
                     ->  Seq Scan on foo_s_1_prt_sub_one  (cost=0.00..961.00 rows=43050 width=4)
               ->  Result  (cost=0.00..961.00 rows=43050 width=4)
                     One-Time Filter: PartSelected
                     ->  Seq Scan on foo_s_1_prt_sub_two  (cost=0.00..961.00 rows=43050 width=4)
         ->  Hash  (cost=2.10..2.10 rows=1 width=4)
               ->  Partition Selector for foo_s (dynamic scan id: 1)  (cost=0.00..2.10 rows=1 width=4)
                     Filter: bar_s.c
                     ->  Broadcast Motion 2:2  (slice2; segments: 2)  (cost=0.00..2.10 rows=1 width=4)
                           ->  Seq Scan on bar_s  (cost=0.00..2.07 rows=1 width=4)
                                 Filter: ((SubPlan 1) = c)
                                 SubPlan 1  (slice2; segments: 2)
                                   ->  Aggregate  (cost=1.04..1.05 rows=1 width=4)
                                         ->  Result  (cost=0.00..1.01 rows=1 width=4)
                                               One-Time Filter: (bar_s.c = 9)
                                               ->  Result  (cost=0.00..1.02 rows=1 width=4)
                                                     ->  Materialize  (cost=0.00..1.02 rows=1 width=4)
                                                           ->  Broadcast Motion 2:2  (slice1; segments: 2)  (cost=0.00..1.01 rows=1 width=4)
                                                                 ->  Seq Scan on baz_s  (cost=0.00..1.01 rows=1 width=4)
                                 SubPlan 1  (slice2; segments: 2)
                                   ->  Aggregate  (cost=1.04..1.05 rows=1 width=4)
                                         ->  Result  (cost=0.00..1.01 rows=1 width=4)
                                               One-Time Filter: (bar_s.c = 9)
                                               ->  Result  (cost=0.00..1.02 rows=1 width=4)
                                                     ->  Materialize  (cost=0.00..1.02 rows=1 width=4)
                                                           ->  Broadcast Motion 2:2  (slice1; segments: 2)  (cost=0.00..1.01 rows=1 width=4)
                                                                 ->  Seq Scan on baz_s  (cost=0.00..1.01 rows=1 width=4)
 Optimizer: Postgres query optimizer
```

I agree with not including the earlier fix until we see the issue. I will spend some time in trying to find a repro if it still exists on your branch.

bhuvnesh2703 commented on src/backend/cdb/cdbllize.c:
```diff
@@ -141,7 +182,13 @@ cdbparallelize(PlannerInfo *root, Plan *plan, Query *query)
 
 	Assert(is_plan_node((Node *) plan));
-	Assert(query != NULL && IsA(query, Query));
+	/* this should only be called at the to -level, not on a subquery */
```
top?

hlinnaka commented:
On 15/09/2019 01:44, Bhuvnesh wrote:
> One of the queries which was added with the commit: d0aea18 
> <https://github.com/greenplum-db/gpdb/commit/d0aea1847518ce86dd0774d3ea20ab967bdba879> 
> produced the plan as below. Two subplans are used once in Hash Cond 
> |Hash Cond: public.foo_s.b = ((subplan))| and other in |Filter: c = 
> ((subplan))|. But now if we see this plan, Hash Cond refers to |Hash 
> Cond: (foo_s_1_prt_sub_one.b = bar_s.c)| where bar_s.c is propagated 
> from the inner child of the HJ so even though in difference slices we 
> don't run into the issue.

That's right. The commit 3ea82ffd1c in this PR is specifically to 
address that case. When constructing join clauses from an equivalence 
class, we now prefer anything that doesn't contain a SubPlan. That 
causes the difference in the Hash Cond in that query.

> bhuvneshchaudhary=# explain SELECT bar_s.c FROM bar_s, foo_s WHERE foo_s.b = (SELECT 
> max(i) FROM baz_s WHERE bar_s.c = 9) AND bar_s.c::int4 = (SELECT max(i) 
> FROM baz_s WHERE bar_s.c = 9);

I just realized that a variant of this still fails:

postgres=#  SELECT bar_s.c FROM bar_s, foo_s WHERE foo_s.b = (SELECT 
max(i) FROM baz_s WHERE bar_s.c = 9) AND foo_s.b = (select bar_s.d::int4);
ERROR:  cannot execute inactive Motion (nodeMotion.c:256)  (seg0 slice3 
127.0.0.1:40000 pid=23874) (nodeMotion.c:256)

It's the same query, but I wrapped the "bar_s.d::int4" in a subquery. 
That brings us back to square one: preferring EC members without 
SubPlans doesn't help when all the members have a SubPlan.

The plan for this failing query looks like this:

postgres=# explain (verbose, costs off) SELECT bar_s.c FROM bar_s, foo_s 
WHERE foo_s.b = (SELECT max(i) FROM baz_s WHERE bar_s.c = 9) AND foo_s.b 
= (select bar_s.d::int4);
                                           QUERY PLAN 

-----------------------------------------------------------------------------------------------
  Gather Motion 3:1  (slice3; segments: 3)
    Output: bar_s.c
    ->  Hash Join
          Output: bar_s.c
          Hash Cond: ((SubPlan 1) = foo_s_1_prt_sub_one.b)
          ->  Broadcast Motion 3:3  (slice2; segments: 3)
                Output: bar_s.c, bar_s.d
                ->  Seq Scan on public.bar_s
                      Output: bar_s.c, bar_s.d
                      Filter: ((SubPlan 1) = (SubPlan 2))
                      SubPlan 1  (slice2; segments: 3)
                        ->  Aggregate
                              Output: max(baz_s.i)
                              ->  Result
                                    Output: baz_s.i
                                    One-Time Filter: (bar_s.c = 9)
                                    ->  Result
                                          Output: baz_s.i
                                          ->  Materialize
                                                Output: baz_s.i
                                                ->  Broadcast Motion 3:3 
  (slice1; segments: 3)
                                                      Output: baz_s.i
                                                      ->  Seq Scan on 
public.baz_s
                                                            Output: baz_s.i
                      SubPlan 2  (slice2; segments: 1)
                        ->  Result
                              Output: (bar_s.d)::integer
          ->  Hash
                Output: foo_s_1_prt_sub_one.b
                ->  Append
                      ->  Seq Scan on public.foo_s_1_prt_sub_one
                            Output: foo_s_1_prt_sub_one.b
                      ->  Seq Scan on public.foo_s_1_prt_sub_two
                            Output: foo_s_1_prt_sub_two.b
          SubPlan 1  (slice3; segments: 3)
            ->  Aggregate
                  Output: max(baz_s.i)
                  ->  Result
                        Output: baz_s.i
                        One-Time Filter: (bar_s.c = 9)
                        ->  Result
                              Output: baz_s.i
                              ->  Materialize
                                    Output: baz_s.i
                                    ->  Broadcast Motion 3:3  (slice1; 
segments: 3)
                                          Output: baz_s.i
                                          ->  Seq Scan on public.baz_s
                                                Output: baz_s.i
  Optimizer: Postgres query optimizer
  Settings: optimizer=off
(50 rows)

I think I know what I would *want* the plan to look like. That is a fine 
plan, but the problem is that we "re-evaluate" the SubPlan in the Hash 
Cond. I'd want the SubPlan result to be bubbled up from the SeqScan to 
the parent, through the Motion, so that the Hash Cond wouldn't need to 
re-evaluate it.

The "set_plan_references" code would handle that, if we put the SubPlan 
result in the target list of the Seq Scan. I'm not sure where the code 
to do that should go to, though. When we build the Plan tree, we don't 
keep track of where SubPlan references are used. It would be pretty 
straightforward detect the situation after the Plan is constructed, but 
it's hard to "fix" the target lists at that stage. Perhaps a new pass 
through the Plan tree, just before "set_plan_references()". I'll give 
that a shot.

- Heikki

guofengrichard commented on src/backend/optimizer/path/equivclass.c:
```diff
@@ -1169,11 +1185,11 @@ generate_join_implied_equalities_normal(PlannerInfo *root,
 					best_eq_op = eq_op;
 					best_score = score;
-					if (best_score == 3)
+					if (best_score == 33)
```
You mean `23` here, don't you?

hlinnaka replied:
Yes, good catch. However, I removed this commit from the PR in the end. With the fix to wrap SubPlans in PlaceHolderVars, a SubPlan is not re-executed if it's use twice in the plan tree, we just bubble up its value from the lower node to the upper node, so there's no reason to avoid using SubPlans here.

guofengrichard commented:
> I think I know what I would *want* the plan to look like. That is a fine plan, but the problem is that we "re-> evaluate" the SubPlan in the Hash Cond. I'd want the SubPlan result to be bubbled up from the SeqScan > to the parent, through the Motion, so that the Hash Cond wouldn't need to re-evaluate it. 

This approach should work. I verified it by making the Motion above the Seq Scan to be Redistribute rather than Broadcast  and get a plan that can work. Because the SubPlan is chosen to be the hash key for the Redistribute, it will be added into the targetlist of Seq Scan on public.bar_s.

The plan looks like:

```
# explain (verbose, costs off) SELECT bar_s.c FROM bar_s, foo_s WHERE foo_s.b = (SELECT
max(i) FROM baz_s WHERE bar_s.c = 9) AND foo_s.b = (select bar_s.d::int4);
                                          QUERY PLAN
-----------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice4; segments: 3)
   Output: bar_s.c
   ->  Hash Join
         Output: bar_s.c
         Hash Cond: (((SubPlan 1)) = foo_s_1_prt_sub_one.b)
         ->  Redistribute Motion 3:3  (slice2; segments: 3)
               Output: bar_s.c, bar_s.d, ((SubPlan 1))
               Hash Key: ((SubPlan 1))
               ->  Seq Scan on public.bar_s
                     Output: bar_s.c, bar_s.d, (SubPlan 1)
                     Filter: ((SubPlan 1) = (SubPlan 2))
                     SubPlan 1  (slice2; segments: 3)
                       ->  Aggregate
                             Output: max(baz_s.i)
                             ->  Result
                                   Output: baz_s.i
                                   One-Time Filter: (bar_s.c = 9)
                                   ->  Result
                                         Output: baz_s.i
                                         ->  Materialize
                                               Output: baz_s.i
                                               ->  Broadcast Motion 3:3  (slice1; segments: 3)
                                                     Output: baz_s.i
                                                     ->  Seq Scan on public.baz_s
                                                           Output: baz_s.i
                     SubPlan 1  (slice2; segments: 3)
                       ->  Aggregate
                             Output: max(baz_s.i)
                             ->  Result
                                   Output: baz_s.i
                                   One-Time Filter: (bar_s.c = 9)
                                   ->  Result
                                         Output: baz_s.i
                                         ->  Materialize
                                               Output: baz_s.i
                                               ->  Broadcast Motion 3:3  (slice1; segments: 3)
                                                     Output: baz_s.i
                                                     ->  Seq Scan on public.baz_s
                                                           Output: baz_s.i
                     SubPlan 2  (slice2; segments: 1)
                       ->  Result
                             Output: (bar_s.d)::integer
         ->  Hash
               Output: foo_s_1_prt_sub_one.b
               ->  Redistribute Motion 3:3  (slice3; segments: 3)
                     Output: foo_s_1_prt_sub_one.b
                     Hash Key: foo_s_1_prt_sub_one.b
                     ->  Append
                           ->  Seq Scan on public.foo_s_1_prt_sub_one
                                 Output: foo_s_1_prt_sub_one.b
                           ->  Seq Scan on public.foo_s_1_prt_sub_two
                                 Output: foo_s_1_prt_sub_two.b
 Optimizer: Postgres query optimizer
(53 rows)
```

hlinnaka commented:
> This approach should work. I verified it by making the Motion above the Seq Scan to be Redistribute rather than Broadcast and get a plan that can work. Because the SubPlan is chosen to be the hash key for the Redistribute, it will be added into the targetlist of Seq Scan on public.bar_s.

Ah, that was a clever trick to test that! Thanks for the testing.

It took me a long time to figure out how to achieve that, but I found a pretty simple solution in the end. The planner already has a mechanism to do pretty much that, called PlaceHolderVars. To be honest, I never fully understood what they're used used for; something to do with outer joins and/or pulled-up subqueries. But what they do is exactly what we want here: force an expression to be evaluated at a specific place in the join tree, and have it propagated upwards from there, instead of re-evaluating it.

So I added a step to the planner, to wrap all SubPlans with PlaceHolderVars. I removed the earlier hack I put in place that preferred non-SubPlans over SubPlans, when deciding which EquivalenceMember to use in a join. Now that we propagate the SubPlan results through the tree, instead of re-evaluating them, a SubPlan shouldn't be any more expensive to evaluate than other expressions, and we don't need it for correctness anymore.

guofengrichard commented:
> It took me a long time to figure out how to achieve that, but I found a pretty simple solution in the end. The planner already has a mechanism to do pretty much that, called PlaceHolderVars. To be honest, I never fully understood what they're used used for; something to do with outer joins and/or pulled-up subqueries. But what they do is exactly what we want here: force an expression to be evaluated at a specific place in the join tree, and have it propagated upwards from there, instead of re-evaluating it.
> 

Brilliant idea to leverage the PlaceHolderVar mechanism. AFAIK PlaceHolderVar is usually used to evaluate expression below an outer join, so that when referencing that same expression above the outer join afterwards, it can avoid evaluating the expression again. Otherwise it may produce wrong result.  For instance:
```
# explain (costs off, verbose) select * from a left join (select i, coalesce(j, 60) from b) sub on a.i = sub.i;
                  QUERY PLAN
----------------------------------------------
 Hash Left Join
   Output: a.i, a.j, b.i, (COALESCE(b.j, 60))
   Hash Cond: (a.i = b.i)
   ->  Seq Scan on public.a
         Output: a.i, a.j
   ->  Hash
         Output: b.i, (COALESCE(b.j, 60))
         ->  Seq Scan on public.b
               Output: b.i, COALESCE(b.j, 60)
(9 rows)
```
The expression `(COALESCE(b.j, 60)` must be evaluated below the outer join. So it will be wrapped in a PlaceHolderVar.

It is very suited here for addressing the subplan re-evaluation issue. Brilliant idea!

However I have two concerns.  One is as said in the comments in `is_simple_subquery`:
```
    /*
     * Don't pull up a subquery that has any volatile functions in its
     * targetlist.  Otherwise we might introduce multiple evaluations of these
     * functions, if they get copied to multiple places in the upper query,
     * leading to surprising results.  (Note: the PlaceHolderVar mechanism
     * doesn't quite guarantee single evaluation; else we could pull up anyway
     * and just wrap such items in PlaceHolderVars ...)
     */
    if (contain_volatile_functions((Node *) subquery->targetList))
        return false;
```
It says "the PlaceHolderVar mechanism doesn't quite guarantee single evaluation". Do you know under what circumstances will it happen?

Another concern is, PlaceHolderVar actually will add constraints to join order, as illustrated in `make_outerjoininfo`.  So if we wrap all SubPlans with PlaceHolderVars, will it prevent us from finding the optimal join order?

hlinnaka commented:
On 24/09/2019 09:21, Richard Guo wrote:
> AFAIK PlaceHolderVar is usually used to evaluate expression below an
> outer join, so that when referencing that same expression above the
> outer join afterwards, it can avoid evaluating the expression again.
> Otherwise it may produce wrong result. For instance:
Ah, thanks for the explanation!

> It says "the PlaceHolderVar mechanism doesn't quite guarantee single 
> evaluation". Do you know under what circumstances will it happen?

Hmm, not sure. Perhaps if it ends up being used in the indexqual on a 
index scan, then it might also be used in the "recheck" of the same 
qual. I think that particular case would be OK for this patch; 
re-evaluating a SubPlan is OK, as long as it's within the same slice.

> Another concern is, PlaceHolderVar actually will add constraints to join 
> order, as illustrated in |make_outerjoininfo|. So if we wrap all 
> SubPlans with PlaceHolderVars, will it prevent us from finding the 
> optimal join order?

Maybe. I'm not sure what the scenarios are, where it constrains the join 
order. But I'm cool with that, our current subplan planning is so 
simplistic that it can only get better.

- Heikki

hlinnaka commented:
Pushed, thanks for the reviews!

hlinnaka commented on src/backend/cdb/cdbllize.c:
```diff
@@ -163,14 +212,25 @@ cdbparallelize(PlannerInfo *root, Plan *plan)
+	context->dispatchParallel = false;
+	context->anyInitPlanParallel = false;
+	context->currentPlanFlow = NULL;
+	context->subplan_info = (PlanProfileSubPlanInfo *)
```
Doesn't really matter, these are just small one-off allocations during planning. We leak a lot of stuff during planning, they get free'd up when the memory context is destroyd.

hlinnaka commented on src/backend/cdb/cdbllize.c:
```diff
@@ -822,87 +958,124 @@ prescan_walker(Node *node, PlanProfile *context)
+		 * tree. That's one way to deal with the issue. However, I haven't
+		 * been able to find a test case that would exhibit that problem. I'm
+		 * not sure if there is some accidental mechanism that prevents
+		 * SubPlans from being propagated across slices, or I just haven't
```
Ah yes, good point. Fixed as a separate commit.
